### DISCUSSION TITLE
Sheet is dismissed and opened again on iOS 15 when scenePhase changes

### DISCUSSION THREAD

**Discussion by Myrronth** [pointfreeco/swift-composable-architecture#2299] [General]
### Description

As soon as the scenePhase environment variable is used in a view that can open multiple sheets using the same @PresentationState property as source, the **second** sheet gets dismissed and is immediately opened again when the scenePhase changes (e.g. dragging down/up the Control Center, or when the bluetooth permission system alert is presented).

This does only occur on iOS 15, on iOS 16 it all works as expected.

- I discovered the bug while on commit 5b0f59533fd7a7dfd3ce55ece38f1f836825323b
- Not tried to reproduce it on the `main` branch as it uses the new navigation tools

### Examples
- TCA-Example with bug: https://github.com/Myrronth/TCASheetDismissalWhenPermissionAlert
- Vanilla-SwiftUI-Example without bug: https://github.com/Myrronth/SheetDismissalWhenPermissionAlert (not sure if this is really the same example only without TCA, but at least it behaves the same way as a user)

### Video that shows the behaviour
https://youtube.com/shorts/Bgx5hEO8CUs

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The sheet stays open, regardless of the scenePhase changes.

### Actual behavior

The sheet gets dismissed and is immediately opened again when the scenePhase changes—but only on iOS 15 (and maybe below, I currently don't have a device/simulator with <iOS 15)

### Steps to reproduce

1. Run the example on a device with iOS 15 (and iOS 16 do see the difference)
2. Tap "Show Feature 1"
3. Drag down the Control Center a bit to put the app in a non-active scenePhase -> Works as expected
4. Tap "Show Next Feature"
5. Drag down the Control Center a bit to put the app in a non-active scenePhase -> Does work as expected on iOS 16, but **not** on iOS 15

### The Composable Architecture version information

f18a72c2dfdc6aecc82e06fed2333bae1cf9a095

### Destination operating system

iOS 15

### Xcode version information

14.3.1

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8.1 (swiftlang-5.8.0.124.5 clang-1403.0.22.11.100)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @Myrronth, unfortunately this is just a vanilla SwiftUI bug. Your vanilla example does not capture it because you use `@State`, but there are many SwiftUI bugs known to plague observable objects that for whatever reason do not affect `@State`.

This recreation of your demo using `@ObservedObject` also shows glitchy behavior:

<details>
<summary>
Click to expand
</summary>

```swift
import SwiftUI

class Model: ObservableObject {
  @Published var showSheet: Bool = false
  @Published var showSheet2: Bool = false
  @Published var shouldShowSheet2: Bool = false
}

struct ContentView: View {
  @Environment(\.scenePhase) var scenePhase
  
  @ObservedObject var model = Model()

  var body: some View {
    VStack {
      Button {
        model.showSheet = true
      } label: {
        Text("Show sheet 1")
      }
    }
    .padding()
    .sheet(isPresented: $model.showSheet, onDismiss: {
      if model.shouldShowSheet2 {
        model.showSheet2 = true
      }
    }) {
      VStack {
        Text("Sheet 1")
        Button {
          model.showSheet = false
          model.shouldShowSheet2 = true
        } label: {
          Text("Show sheet 2")
        }
      }
    }
    .sheet(isPresented: $model.showSheet2) {
      Text("Sheet 2")
    }
  }
}

struct ContentView_Previews: PreviewProvider {
  static var previews: some View {
    ContentView()
  }
}
```    
</details>

Maybe there is some kind of workaround to get fix the behavior, in TCA or vanilla SwiftUI, but as of right now it does not seem like a bug that the library itself can fix. As such I am going to convert this to a discussion.

**Reply by Myrronth** [reply]
Thanks for the quick reply! Totally forgot about @ObservedObject and its bugs. We worked around it by using a dependency that listens for `didBecomeActiveNotification` and the like (we don't support Scene’s), iterate over its AsyncStream in the reducers that need it, and removed all scenePhase environment variables.

**Reply by Myrronth** [reply]
We encountered a similar bug in another case where a sheet was presented after a button in a previous sheet was tapped. Both were driven of the same `destination` property and on iOS 16 it just worked and the transition looked fine (first closing a sheet, then opening the other one).

But every once in a while, we got the known "cannot present sheet as another sheet is already presented" (forgot the actual error message) error on iOS 16. To fix that, we refactored it to delay setting the new destination to when the `onDismiss` callback from TCA’s sheet modifier is called instead of when a button in the first sheet was tapped.

Turns out, delaying the destination update also fixes the initial bug mentioned at the start of this discussion and also the bug mentioned at the start of this reply.

**Reply by mbrandonw** [reply]
Good find! Yeah this is a known bug in SwiftUI and even plagues UIKit sometimes. Simultaneously performing a dismissal and navigation can mix up the framework. It's unfortunate, but luckily there are workarounds.
